### PR TITLE
basalt-monado: init at 0-unstable-2024-06-21

### LINKED ISSUE #294764: Package request: Basalt for Monado
**Project description**
[Basalt for Monado](https://gitlab.freedesktop.org/mateosss/basalt) is a fork of [Basalt](https://gitlab.com/VladyslavUsenko/basalt), the implementation of a [Visual-Inertial Mapping with Non-Linear Factor Recovery](https://vision.in.tum.de/research/vslam/basalt) algorithm, for use on [Monado](https://monado.freedesktop.org/), as mentioned in https://github.com/NixOS/nixpkgs/pull/245005.

There are some dependencies that need to be on nixpkgs for basalt to be packaged:
- [x] [Sophus](https://github.com/strasdat/Sophus) - https://github.com/NixOS/nixpkgs/pull/294742
- [x] [OpenGV](https://github.com/laurentkneip/opengv) - https://github.com/NixOS/nixpkgs/pull/293265
- [x] ~[Pangolin v0.6](https://github.com/stevenlovegrove/Pangolin/tree/v0.6) - https://github.com/NixOS/nixpkgs/pull/293337~ Overriden pangolin with patches

**Metadata**

* homepage URL: https://gitlab.freedesktop.org/mateosss/basalt
* source URL: https://gitlab.freedesktop.org/mateosss/basalt
* license: bsd3
* platforms: linux

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by loco-choco**
## Description of changes
[Basalt for Monado](https://gitlab.freedesktop.org/mateosss/basalt) "is a fork of [Basalt](https://gitlab.com/VladyslavUsenko/basalt) improved for tracking XR devices with [Monado](https://gitlab.freedesktop.org/monado/monado)."
Adds the package in https://github.com/NixOS/nixpkgs/issues/294764 for https://github.com/NixOS/nixpkgs/pull/245005#issuecomment-1962320314.

Closes NixOS/nixpkgs#294764

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by loco-choco** [comment]
For testing the executables, try the example on [Usage](https://gitlab.freedesktop.org/mateosss/basalt#usage) on the [Basalt for Monado](https://gitlab.freedesktop.org/mateosss/basalt) repository.

File: pkgs/by-name/ba/basalt-monado/package.nix [hunk 1]
```diff
@@ -0,0 +1,112 @@
+{ stdenv,
+  lib,
+  boost,
+  eigen,
+  tbb_2021_8,
```
Comment (Scrumplex) [inline pkgs/by-name/ba/basalt-monado/package.nix:5]: With versioned packages like these, you should take `tbb` as an input and then manually `callPackage` this derivation in `top-level/all-packages.nix` with the corresponding version

This is currently one of the shortcomings of [RFC-140](https://github.com/NixOS/rfcs/blob/master/rfcs/0140-simple-package-paths.md)

File: pkgs/by-name/ba/basalt-monado/package.nix [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+  fmt,
+  bzip2,
+  opengv,
+  cli11,
+  magic-enum,
+  sophus,
+  cereal_1_3_2,
```
Comment (Scrumplex) [inline pkgs/by-name/ba/basalt-monado/package.nix:12]: this as well

File: pkgs/by-name/ba/basalt-monado/package.nix [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    bzip2
+    opengv
+    cli11
+    magic-enum
+    sophus
+    cereal_1_3_2
+    freeglut
+    glew
+    libGL
+    (opencv.override{ enableGtk3 = true; })
```
Comment (Scrumplex) [inline pkgs/by-name/ba/basalt-monado/package.nix:90]: this should also be moved to a callPackage call

**Comment by Scrumplex** [comment]
Also, this doesn't build when rebased onto master, because `tbb_2021_8` doesn't exist anymore: https://github.com/NixOS/nixpkgs/pull/301391

**Comment by loco-choco** [comment]
> Also, this doesn't build when rebased onto master, because `tbb_2021_8` doesn't exist anymore: #301391

Updating to `tbb_2021_11` seems fine, so I will just change it to that.

**Comment by Scrumplex** [comment]
Result of `nixpkgs-review pr 295107` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>basalt-monado</li>
  </ul>
</details>

File: pkgs/applications/graphics/basalt-monado/default.nix
Comment (Scrumplex) [inline pkgs/applications/graphics/basalt-monado/default.nix:1]: Oh I should have been a little more precise. So we can keep this package in `pkgs/by-name` we just add a callPackage to `all-packages.nix` that will override the automatic invocation.

See this package as an example: https://github.com/NixOS/nixpkgs/pull/302517

Comment (loco-choco) [reply pkgs/applications/graphics/basalt-monado/default.nix:1]: Ah, sorry, my bad. It should be fine now.

**Review by Scrumplex — APPROVED** [review]
Derivation looks good to me. Haven't had an opportunity to test this though

**Comment by Scrumplex** [comment]
Result of `nixpkgs-review pr 295107` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>basalt-monado</li>
  </ul>
</details>

**Review by Scrumplex — APPROVED** [review]

**Review by Pandapip1 — APPROVED** [review]

File: pkgs/top-level/all-packages.nix [hunk 4]
```diff
@@ -438,6 +438,12 @@ with pkgs;
 
   catppuccin-plymouth = callPackage ../data/themes/catppuccin-plymouth { };
 
+  basalt-monado = callPackage ../by-name/ba/basalt-monado/package.nix {
+    tbb = tbb_2021_11;
+    cereal = cereal_1_3_2;
+    opencv = opencv.override{ enableGtk3 = true; };
+  };
+
```
Comment (Pandapip1) [inline pkgs/top-level/all-packages.nix:446]:
```suggestion
```
Wait whoops, that shouldn't be there.

**Review by Pandapip1 — CHANGES_REQUESTED** [review]

File: pkgs/top-level/all-packages.nix (hunk 4 above)
Comment (Pandapip1) [inline pkgs/top-level/all-packages.nix:446]: Packages in `by-name` are included by default without having to add them to `all-packages.nix`.

I'm pretty sure you should be overriding opencv in the derivation, *NOT* here.

**Review by Pandapip1 — CHANGES_REQUESTED** [review]
Almost missed something

File: pkgs/by-name/ba/basalt-monado/package.nix [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+  pangolin,
+  fetchFromGitLab,
+  fetchFromGitHub,
+  cmake,
+  pkg-config,
+  autoPatchelfHook,
+  extra-cmake-modules,
+  nix-update-script
+}:
+stdenv.mkDerivation (finalAttrs: {
```
Comment (Pandapip1) [inline pkgs/by-name/ba/basalt-monado/package.nix:26]:
```suggestion
stdenv.mkDerivation {
```

File: pkgs/by-name/ba/basalt-monado/package.nix [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+  meta = {
+    description = "A fork of Basalt improved for tracking XR devices with Monado";
+    homepage = "https://gitlab.freedesktop.org/mateosss/basalt";
+    license = lib.licenses.bsd3;
+    mainProgram = "basalt_vio";
+    maintainers = [ lib.maintainers.locochoco ];
+    platforms = lib.platforms.all;
+  };
+})
```
Comment (Pandapip1) [inline pkgs/by-name/ba/basalt-monado/package.nix:112]:
```suggestion
}
```

File: pkgs/by-name/ba/basalt-monado/package.nix [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    cli11
+    magic-enum
+    sophus
+    cereal
+    freeglut
+    glew
+    libGL
+    opencv
+    finalAttrs.pangolin_0_6
+  ];
```
Comment (Pandapip1) [inline pkgs/by-name/ba/basalt-monado/package.nix:92]:
```suggestion
  nativeBuildInputs = [
    cmake
    cli11
    extra-cmake-modules
    pkg-config
    autoPatchelfHook
  ];

  buildInputs = [
    boost
    eigen
    tbb
    fmt
    bzip2
    opengv
    cli11
    magic-enum
    sophus
    cereal
    freeglut
    glew
    libGL
    (opencv.override {
      enableGtk3 = true;
    })
    (pangolin.overrideAttrs {
      version = "0.6";
      src = fetchFromGitHub {
        owner = "stevenlovegrove";
        repo = "Pangolin";
        rev = "86eb4975fc4fc8b5d92148c2e370045ae9bf9f5d";
        sha256 = "sha256-fKteOuOuGMWPZFxOUGCUcjeLXtTUXSGMSs1QfM5qblU=";
      };
      patches = [ ./pangolin-0_6-cstdint.patch ];
    })
  ];
```

File: pkgs/by-name/ba/basalt-monado/package.nix [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+  opencv,
+  pangolin,
+  fetchFromGitLab,
+  fetchFromGitHub,
+  cmake,
+  pkg-config,
+  autoPatchelfHook,
+  extra-cmake-modules,
+  nix-update-script
+}:
```
Comment (Pandapip1) [inline pkgs/by-name/ba/basalt-monado/package.nix:25]: This list is out of order.
```suggestion
{ stdenv,
  lib,
  fetchFromGitLab,
  fetchFromGitHub,
  autoPatchelfHook,
  boost,
  bzip2,
  cereal,
  cli11,
  cmake,
  eigen,
  extra-cmake-modules,
  fmt,
  freeglut,
  glew,
  libGL,
  magic-enum,
  opencv,
  opengv,
  pangolin,
  pkg-config,
  sophus,
  tbb,
  nix-update-script
}:
```

File: pkgs/top-level/all-packages.nix (hunk 4 above)
Comment (Pandapip1) [inline pkgs/top-level/all-packages.nix:446]: Packages in `by-name` are included by default without having to add them to `all-packages.nix`.

You should be overriding opencv in the derivation, NOT here.

```suggestion
```

**Review by Pandapip1 — CHANGES_REQUESTED** [review]

File: pkgs/by-name/ba/basalt-monado/package.nix [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+  passthru.updateScript = nix-update-script {};
+
+  meta = {
+    description = "A fork of Basalt improved for tracking XR devices with Monado";
+    homepage = "https://gitlab.freedesktop.org/mateosss/basalt";
+    license = lib.licenses.bsd3;
+    mainProgram = "basalt_vio";
+    maintainers = [ lib.maintainers.locochoco ];
+    platforms = lib.platforms.all;
+  };
```
Comment (Pandapip1) [inline pkgs/by-name/ba/basalt-monado/package.nix:111]: Typically `meta` is formatted like so:

```suggestion
  meta = with lib; {
    description = "A fork of Basalt improved for tracking XR devices with Monado";
    homepage = "https://gitlab.freedesktop.org/mateosss/basalt";
    license = licenses.bsd3;
    mainProgram = "basalt_vio";
    maintainers = with maintainers; [ locochoco ];
    platforms = platforms.all;
  };
```

**Review by Pandapip1 — CHANGES_REQUESTED** [review]

File: pkgs/by-name/ba/basalt-monado/package.nix [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+  fetchFromGitLab,
+  fetchFromGitHub,
+  cmake,
+  pkg-config,
+  autoPatchelfHook,
+  extra-cmake-modules,
+  nix-update-script
+}:
+stdenv.mkDerivation (finalAttrs: {
+  pname = "basalt";
```
Comment (Pandapip1) [inline pkgs/by-name/ba/basalt-monado/package.nix:27]:
```suggestion
  pname = "basalt-monado";
```

File: pkgs/top-level/all-packages.nix (hunk 4 above)
Comment (Scrumplex) [reply pkgs/top-level/all-packages.nix:446]: actually, for simple overrides like this it is quite common to place them in `pkgs/top-level/all-packages.nix`. Same goes for the other versioned dependencies

File: pkgs/by-name/ba/basalt-monado/package.nix (hunk 7 above)
Comment (loco-choco) [reply pkgs/by-name/ba/basalt-monado/package.nix]: Should be fixed now.

File: pkgs/by-name/ba/basalt-monado/package.nix (hunk 8 above)
Comment (loco-choco) [reply pkgs/by-name/ba/basalt-monado/package.nix]: Also fixed now.

File: pkgs/by-name/ba/basalt-monado/package.nix (hunk 10 above)
Comment (loco-choco) [reply pkgs/by-name/ba/basalt-monado/package.nix]: Yeah, makes sense, fixed now.

File: pkgs/by-name/ba/basalt-monado/package.nix (hunk 5 above)
Comment (Pandapip1) [reply pkgs/by-name/ba/basalt-monado/package.nix:26]: Got it. This can be marked as resolved.

File: pkgs/top-level/all-packages.nix (hunk 4 above)
Comment (Pandapip1) [reply pkgs/top-level/all-packages.nix:446]: Fair enough.

**Review by Pandapip1 — APPROVED** [review]

**Comment by BaconCondensate** [comment]
Nix/flake noob here, idk if this is the correct place to ask.  But how do you build this with monado?
doing `nix shell github:nixos/nixpkgs/pull/295107/head#basalt-monado` and then 
`basalt_vio --show-gui 1 --dataset-path MGO09_short_1_updown --dataset-type euroc --cam-calib /nix/store/m07mkqr8br9qpp3gcx75yk27i50m7xi2-basalt-monado-release-673cc5c6/share/basalt/msdmi_calib.json --config-path /nix/store/m07mkqr8br9qpp3gcx75yk27i50m7xi2-basalt-monado-release-673cc5c6/share/basalt/msdmi_config.json `

seemed to work for testing, but getting it to work with monado (eg. `monado-cli slambatch MGO09_short_1_updown/`) failed and resulted in `No SLAM system built.`

**Comment by loco-choco** [comment]
I think you need to either compile monado with basalt, so you would need to override the derivation to add the package to the build inputs, or set the env var `VIT_SYSTEM_LIBRARY_PATH` as described here <https://gitlab.freedesktop.org/mateosss/basalt#usage>.

**Comment by BaconCondensate** [comment]
> I think you need to either compile monado with basalt, so you would need to override the derivation to add the package to the build inputs, or set the env var `VIT_SYSTEM_LIBRARY_PATH` as described here https://gitlab.freedesktop.org/mateosss/basalt#usage.

How do you compile monado with basalt on nix? Just setting `VIT_SYSTEM_LIBRARY_PATH` via `VIT_SYSTEM_LIBRARY_PATH=/nix/store/m07mkqr8br9qpp3gcx75yk27i50m7xi2-basalt-monado-release-673cc5c6/lib/libbasalt.so` doesn't seem to work

currently I install monado via this nix file

```
{
  inputs,
  lib,
  config,
  pkgs,
  ...
}: {

  services.monado = {
    enable = true;
    defaultRuntime = true; # Register as default OpenXR runtime
  };

  systemd.user.services.monado.environment = {
    STEAMVR_LH_ENABLE = "1";
    XRT_COMPOSITOR_COMPUTE = "1";
    WMR_HANDTRACKING= "FALSE";
  };

  environment.systemPackages = with pkgs; [
    onnxruntime
    libusb1
    hidapi
    opencomposite
    wlx-overlay-s
    xrgears
    openxr-loader
    #basalt-monado #check https://github.com/NixOS/nixpkgs/pull/295107/

    #envision
  ];
}
```

**Comment by BaconCondensate** [comment]
Any update to this getting merged yet? Seems like it's waiting review from @drupol ?

**Comment by drupol** [comment]
I'll pass this time, I don't have time today sorry :(

File: pkgs/top-level/all-packages.nix [hunk 11]
```diff
@@ -326,6 +326,12 @@ with pkgs;
 
   banana-accounting = callPackage ../applications/office/banana-accounting { };
 
+  basalt-monado = callPackage ../by-name/ba/basalt-monado/package.nix {
+    tbb = tbb_2021_11;
+    cereal = cereal_1_3_2;
+    opencv = opencv.override{ enableGtk3 = true; };
+  };
+
```
Comment (Pandapip1) [inline pkgs/top-level/all-packages.nix:334]:
```suggestion
```
Files in `pkgs/by-name` are automatically called

File: pkgs/by-name/ba/basalt-monado/fixed-cmakes.patch [hunk 12]
```diff
@@ -0,0 +1,353 @@
...
++#    set(_include_dirs_new "")
++#    foreach(_dir IN LISTS _include_dirs)
++#      if(_dir MATCHES ".*/eigen(/unsupported)?$")
++#        string(REGEX REPLACE "(^${CMAKE_SOURCE_DIR}.*$)" "$<BUILD_INTERFACE:\\1>" _dir "${_dir}")
++#      endif()
++#      list(APPEND _include_dirs_new "${_dir}")
++#    endforeach()
++#    set_target_properties(${_target} PROPERTIES INTERFACE_INCLUDE_DIRECTORIES "${_include_dirs_new}")
++#  endforeach()
++#endif()
```
Comment (Pandapip1) [inline pkgs/by-name/ba/basalt-monado/fixed-cmakes.patch:339]: Any reason this is commented out instead of removed?

**Review by Pandapip1 — APPROVED** [review]
LGTM

File: pkgs/by-name/ba/basalt-monado/package.nix [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+  passthru.updateScript = nix-update-script {};
+
+  meta = with lib; {
+    description = "A fork of Basalt improved for tracking XR devices with Monado";
+    homepage = "https://gitlab.freedesktop.org/mateosss/basalt";
+    license = licenses.bsd3;
+    mainProgram = "basalt_vio";
+    maintainers = [ maintainers.locochoco ];
+    platforms = platforms.all;
+  };
```
Comment (Pandapip1) [inline pkgs/by-name/ba/basalt-monado/package.nix:109]: Nit: incorrect `platforms` and remove `with lib`.
```suggestion
  meta = {
    description = "A fork of Basalt improved for tracking XR devices with Monado";
    homepage = "https://gitlab.freedesktop.org/mateosss/basalt";
    license = lib.licenses.bsd3;
    mainProgram = "basalt_vio";
    maintainers = with lib.maintainers; [ locochoco ];
    platforms = lib.platforms.linux;
  };
```

File: pkgs/by-name/ba/basalt-monado/package.nix [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+  opengv,
+  pangolin,
+  pkg-config,
+  sophus,
+  tbb,
+  nix-update-script
+}:
+stdenv.mkDerivation (finalAttrs: {
+  pname = "basalt-monado";
+  version = "release-673cc5c6";
```
Comment (Pandapip1) [inline pkgs/by-name/ba/basalt-monado/package.nix:28]:
```suggestion
  version = "0-release-673cc5c6";
```
Versions must start with a number: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md?plain=1#L350

File: pkgs/top-level/all-packages.nix [hunk 15]
```diff
@@ -326,6 +326,12 @@ with pkgs;
 
   banana-accounting = callPackage ../applications/office/banana-accounting { };
 
+  basalt-monado = callPackage ../by-name/ba/basalt-monado/package.nix {
+    tbb = tbb_2021_11;
+    cereal = cereal_1_3_2;
+    opencv = opencv.override{ enableGtk3 = true; };
```
Comment (Pandapip1) [inline pkgs/top-level/all-packages.nix:332]: I'm not certain, but I don't think this follows https://search.nixos.org/packages?show=nixfmt-rfc-style

**Review by Pandapip1 — CHANGES_REQUESTED** [review]
Nit: If any of the patches have been submitted as MRs, it makes sense to pull them using `fetchpatch`.

Also, format the files using nixfmt-rfc-style

File: pkgs/top-level/all-packages.nix (hunk 15 above)
Comment (SuperSandro2000) [reply pkgs/top-level/all-packages.nix:332]:
```suggestion
    opencv = opencv.override { enableGtk3 = true; };
```

File: pkgs/by-name/ba/basalt-monado/package.nix (hunk 14 above)
Comment (SuperSandro2000) [reply pkgs/by-name/ba/basalt-monado/package.nix:28]: That's the release tag upstream uses https://gitlab.freedesktop.org/mateosss/basalt/-/tags

but we would strip the release from it.

Comment (eclairevoyant) [reply pkgs/by-name/ba/basalt-monado/package.nix:28]: This is not really a meaningful version IMO.

File: pkgs/by-name/ba/basalt-monado/package.nix (hunk 13 above)
Comment (loco-choco) [reply pkgs/by-name/ba/basalt-monado/package.nix]: The patches weren't submitted as MR's, maybe in the future they can be reworked to work as such.

**Comment by BaconCondensate** [comment]
any updates to this yet?

File: pkgs/by-name/ba/basalt-monado/package.nix [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+  };
+
+  postUnpack = ''
+    rm -r source/thirdparty/basalt-headers
+    ln -s ${finalAttrs.headers-src} source/thirdparty/basalt-headers
+  '';
+
+  patches = [
+    ./fixed-cmakes.patch
+    ./remove-ros.patch
```
Comment (SuperSandro2000) [inline pkgs/by-name/ba/basalt-monado/package.nix:54]: That patch doesn't apply

File: pkgs/by-name/ba/basalt-monado/package.nix [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+    (pangolin.overrideAttrs {
+      version = "0.6";
+      src = fetchFromGitHub {
+        owner = "stevenlovegrove";
+        repo = "Pangolin";
+        rev = "86eb4975fc4fc8b5d92148c2e370045ae9bf9f5d";
+        sha256 = "sha256-fKteOuOuGMWPZFxOUGCUcjeLXtTUXSGMSs1QfM5qblU=";
+      };
+      patches = [ ./pangolin-0_6-cstdint.patch ];
+    })
```
Comment (SuperSandro2000) [inline pkgs/by-name/ba/basalt-monado/package.nix:90]: We should move that into a let in and name it pangolin_0_6 or pangolin'.

The reason is, that inlining such a big thing is clunky and we can't reuse the package.

File: pkgs/by-name/ba/basalt-monado/package.nix (hunk 14 above)
Comment (eclairevoyant) [reply pkgs/by-name/ba/basalt-monado/package.nix:29]: why is this not being fixed? this is a nonsensical version.

Comment (Pandapip1) [reply pkgs/by-name/ba/basalt-monado/package.nix:29]: Because:

a) Versions must start with a digit
b) This is not an "unstable" version; this is actually a release
c) The short commit identifier of the release is `673cc5c6`.

Comment (eclairevoyant) [reply pkgs/by-name/ba/basalt-monado/package.nix:29]: > a) Versions must start with a digit

Which this clearly violates.


> c) The name of the git tag is `release-673cc5c6`.

And again, this is not meaningful, it's just some rando git commit, that's not a numerical version.

Comment (Pandapip1) [reply pkgs/by-name/ba/basalt-monado/package.nix:29]: > Which this clearly violates.

Oh, I thought they'd applied my patch. Yea, that needs fixing.

> And again, this is not meaningful, it's just some rando git commit, that's not a numerical version.

There are no numerical versions to basalt-monado. According to the applicable [naming guidelines](https://github.com/NixOS/nixpkgs/blob/ff93fa8385374ab39eafc2dc882f699bc010f49a/pkgs/README.md?plain=1#L344-L362)

- The `pname` and the `version` attribute _must not_ contain uppercase letters — e.g., `"mplayer"` instead of `"MPlayer"`.

Satisfied.

- The `version` attribute _must_ start with a digit e.g., `"0.3.1rc2"` or `"0-unstable-1970-01-01"`.

Not satisfied, ergo my patch.

 - If a package is a commit from a repository without a version assigned, then the `version` attribute _should_ be the latest upstream version preceding that commit, followed by `-unstable-` and the date of the (fetched) commit. The date _must_ be in `"YYYY-MM-DD"` format.

Not applicable, as there is an assigned version.

- If a project has no suitable preceding releases - e.g., no versions at all, or an incompatible versioning / tagging schema - then the latest upstream version in the above schema should be `0`.

Not applicable, as there are releases.

So, in summary: the only blocking change is to prefix the release with a digit.

File: pkgs/by-name/ba/basalt-monado/package.nix (hunk 16 above)
Comment (loco-choco) [reply pkgs/by-name/ba/basalt-monado/package.nix]: I was looking more into the cmake file and found out that there weren't really needed, so now the patches are removed, but now it compiles some of its dependencies from a git submodule. I tested the main binary and everything seems fine, but if we _must_ build the third party dependencies as derivations, I can look into making these patches look better.

File: pkgs/by-name/ba/basalt-monado/package.nix (hunk 17 above)
Comment (loco-choco) [reply pkgs/by-name/ba/basalt-monado/package.nix]: Removed it, for the reasons in https://github.com/NixOS/nixpkgs/pull/295107#discussion_r1720542684.

File: pkgs/by-name/ba/basalt-monado/package.nix (hunk 14 above)
Comment (loco-choco) [reply pkgs/by-name/ba/basalt-monado/package.nix]: Sorry for taking so long to say decide on how to handle this, I was waiting for what the consensus would be for it and only now had time to properly see this, I decided to go with this part of the guideline:
> If a project has no suitable preceding releases - e.g., no versions at all, or an incompatible versioning / tagging schema - then the latest upstream version in the above schema should be `0`.
> 
> Example: Given a project that has no tags / released versions at all, or applies versionless tags like `latest` or `YYYY-MM-DD-Build`, a commit authored on March 15, 2022 would have `version = "0-unstable-2022-03-15"`. 

So the version is now `0-unstable-2024-06-21`.

Comment (loco-choco) [reply pkgs/by-name/ba/basalt-monado/package.nix]: I also upgraded to the latest "release" on the repo.

**Review by Scrumplex — APPROVED** [review]

**Review by Pandapip1 — APPROVED** [review]
I believe all my concerns were addressed.
